### ISSUE TITLE
Unnecessary 2FA Prompt and notification email

### ISSUE THREAD

**Issue by mckinleytabor** [MauriceConrad/iCloud-API#7]
Just wondering if this is happening to anyone else. I've gotten everything working, and my scripts execute correctly without any errors.

However, every time it runs, I get a prompt on my Apple Devices (iPhone, Mac) to do 2FA, and an email indicating that a new devices has signed into iCloud. Obviously this was necessary the first time to input the code, but these subsequent requests seems to have no effect on the script running. I.e. it runs correctly regardless of what I do with the prompt, (ignore, dismiss, etc)

I believe that I am using my session JSON file correctly, otherwise I doubt I would be able to access items like Reminders and Find my Friends without having to reauthenticate. (my scripts do not have my AppleID or passwords in them, these are only in the JSON)

Any insight on what might be causing these 2FA prompts and notification emails, or is this just something with Apple that is unavoidable?

**Comment by Lawris** [comment]
Same here, if anyone has an explanation and how to stop those annoying e-mails...

**Comment by joshstrange** [comment]
I just ran into this today. It appears to be due to this cookie:

```
      {
        "X-APPLE-WEBAUTH-HSA-LOGIN": "",
        "Expires": "Thu, 01-Jan-1970 00:00:01 GMT",
        "Path": "/",
        "Domain": ".icloud.com"
      },
```

There is code in the library that checks if ANY cookies are expired and if they are it will try to re-login thus the 2FA request. (second part of the sentence has not been confirm by looking at code but experience).

If I delete that cookie out of the session my script doesn't throw the error:

```
{ error: 'Session is expired or invalid', errorCode: 6 }
```

Which leads to the 2FA prompt. I have not tested yet but I would assume it re-adds it on each session save (assuming Apple keeps sending that cookie back). I'm still considering which approach is best:

1. Fork lib and filter cookies that are in the past of now at time of saving session so that obviously invalid cookies won't be saved in the first place

2. Fork lib and ignore this specific cookie

3. In sessionUpdate hook delete the cookie in the saved json file myself using either method above

4. Using some other hook delete the cookie out of the running library and rely on sessionUpdate firing on to save the fix to file

**Comment by joshstrange** [comment]
After testing some more my original assumption may not have been correct. I've having trouble getting my code into a state that consistently works without email/2FA prompt. I still think errorCode: 6 is probably what is kicking off the email (or is fired by code that then hits an endpoint causing the prompt.

**Comment by vascobarbosa** [comment]
I do belive it because of this :

"hsaTrustedBrowser": false

in session.json

**Comment by vascobarbosa** [comment]
Icloud does exactly the same thing.

expire date unix time 0 
<img width="534" alt="Screen Shot 2019-06-23 at 2 48 37 PM" src="https://user-images.githubusercontent.com/2828849/59977194-01eec800-95c6-11e9-965f-14c67548f4eb.png">

**Comment by Tobix99** [comment]
I'm also very interested in this feature. Is there any chance to not get the notification?

**Comment by Tobix99** [comment]
So I've fixed the issue where the promt is coming at the first time when the app is starting --> just ignoring the cookie date @vascobarbosa found, because it's alsways set to 1970

but I think I can't get around the restriction, that when I trigger a FindMe request the promt is still popping up

**Comment by christianrexwinkel** [comment]
I think I figured out the problem and the solution.

At line 133 of main.js the loggedIn flag is evaluated:
`self.loggedIn = (function() {
        return (self.auth.cookies.length > 0 && self.auth.token && self.account != {} && self.username === username);
})();`

In my case I was using the prompt-credentials.js file from the demo folder, which creates an iCloud instance by just providing a path to the session file. From line 35:
`var myCloud = new iCloud(sessionPath);`

As a result the loggedIn flag evaluates always to false as self.username (username string from the previous session) != username (empty string from constructor).

By including a username in the iCloud instance creation, I no longer have a 2FA prompt.
`var myCloud = new iCloud(sessionPath, previously_used_username);`

Including this makes sense as the cookie must be evaluated against the intended user login.

**Comment by Tobix99** [comment]
@christianrexwinkel are you sure, that you don't get a promt again after like 30 minutes? Because I always pass my email address along with my pw and the session.js

Btw if you look in my pullrequest I also found the invalid cookie (so `cookiesExpired` in the next step at `self.cookiesValid = (function() {...` after your statement was alsways true) and tried to filter it....it solves it for the first login, but not for the next times

Aaaaand the session will sometimes break, but I think thats a bug from apple, because with the original code I also got an Invalid session after like 30 minutes or so

**Comment by christianrexwinkel** [comment]
Yes I don't get a prompt after 30 minutes, not even after a full day. This morning it still worked. I forgot to mention that I also included your pull request.

Are you sure you're saving your session upon a successful logon?
`myCloud.on("sessionUpdate", function() {
            myCloud.saveSession();
});`

**Comment by Tobix99** [comment]
Ahh, I see the Problem ^^

I wanted to use it to track my devices, so I tried to use the FindMe function every 30 minutes, but then gave me a session invalid error. I will try to update my pull request and maybe open a new for your changes this weekend.

**Comment by bland328** [comment]
@christianrexwinkel, I'm also using the `prompt-credentials.js` file, and have a valid path in `sessionPath`, as I see the session file is created.

I've changed line 35 like this...

`var myCloud = new iCloud(sessionPath, "my@email.here");`

...but I still get a fresh 2FA prompt every time I run a demo script (I've been experimenting with `reminders/getCompletedTasks.js`).

Also, the email address I've provided absolutely matches the email address from the session file. And I've also wiped my session file and started fresh, but no change.

Any thoughts on what I might be getting wrong?
